File: examples/tree_example.cxx
#include <memory>
#include <chrono>
#include <iostream>
#include <sstream>
#include <functional>

#include <ROOT/RVec.hxx>
#include "TPad.h"
#include "TFile.h"
#include "TLorentzVector.h"
#include "TTreeReaderValue.h"
#include "TTreeReader.h"
#include "TVector2.h"
#include "TH1F.h"

#include "ana/analysis.h"

#include "rootana/Tree.h"
#include "rootana/Hist.h"
#include "rootana/Folder.h"

using cut = ana::selection::cut;
using weight = ana::selection::weight;

using VecF = ROOT::RVec<float>;
using VecD = ROOT::RVec<double>;
using VecI = ROOT::RVec<int>;
using VecUI = ROOT::RVec<unsigned int>;
using P4 = TLorentzVector;

class NthP4 : public ana::column::definition<P4(VecD, VecD, VecD, VecD)>
{

public:
  NthP4(unsigned int index) : 
    m_index(index)
  {}
  virtual ~NthP4() = default;

  virtual P4 evaluate(ana::observable<VecD> pt, ana::observable<VecD> eta, ana::observable<VecD> phi, ana::observable<VecD> es) const override {
    P4 p4;
    p4.SetPtEtaPhiE(pt->at(m_index),eta->at(m_index),phi->at(m_index),es->at(m_index));
    return p4;
  }

protected:
  const unsigned int m_index;

};
 
int main() {

  ana::dataflow<Tree> df( {"hww.root"}, "mini" );

  auto mc_weight = df.read<float>("mcWeight");
  auto mu_sf = df.read<float>("scaleFactor_MUON");

  // auto el_sf = df.read<float>("scaleFactor_ELE");
  auto el_sf = df.read<float>("scaleFactor_ELE").vary("sf_var","scaleFactor_PILEUP");

  auto lep_pt_MeV = df.read<VecF>("lep_pt");
  auto lep_eta = df.read<VecF>("lep_eta");
  auto lep_phi = df.read<VecF>("lep_phi");
  auto lep_E_MeV = df.read<VecF>("lep_E");
  auto lep_Q = df.read<VecF>("lep_charge");
  auto lep_type = df.read<VecUI>("lep_type");
  auto met_MeV = df.read<float>("met_et");
  auto met_phi = df.read<float>("met_phi");

  auto MeV = df.constant(1000.0);
  auto lep_pt = lep_pt_MeV / MeV;
  auto lep_E = lep_E_MeV / MeV;
  auto met = met_MeV / MeV;

  auto lep_eta_max = df.constant(2.4);

  // auto Escale = df.define([](VecD E){return E;});
  // auto Escale = df.define([](VecD E){return E;}).vary("lp4_up",[](VecD E){return E*1.01;});
  auto Escale = df.define([](VecD E){return E;}).vary("lp4_up",[](VecD E){return E*1.01;}).vary("lp4_dn",[](VecD E){return E*0.99;});

  auto lep_pt_sel = Escale(lep_pt)[ lep_eta < lep_eta_max && lep_eta > (-lep_eta_max) ];
  auto lep_E_sel = Escale(lep_E)[ lep_eta < lep_eta_max && lep_eta > (-lep_eta_max) ];
  auto nlep_sel = df.define([](VecD const& lep){return lep.size();})(lep_pt_sel);

  std::cout << nlep_sel.has_variation("lp4_up") << std::endl;

  auto lep_eta_sel = lep_eta[ lep_eta < lep_eta_max && lep_eta > (-lep_eta_max) ];
  auto lep_phi_sel = lep_phi[ lep_eta < lep_eta_max && lep_eta > (-lep_eta_max) ];

  auto l1p4 = df.define<NthP4>(0)(lep_pt_sel, lep_eta_sel, lep_phi_sel, lep_E_sel);
  auto l2p4 = df.define<NthP4>(1)(lep_pt_sel, lep_eta_sel, lep_phi_sel, lep_E_sel);

  auto llp4 = l1p4+l2p4;
  auto mll = df.define([](const P4& p4){return p4.M();})(llp4);
  auto pth = df.define(
    [](const P4& p4, float q, float q_phi) {
      TVector2 p2; p2.SetMagPhi(p4.Pt(), p4.Phi());
      TVector2 q2; q2.SetMagPhi(q, q_phi);
      return (p2+q2).Mod();
    })(llp4, met, met_phi);

  auto incl = df.filter<weight>("incl")(mc_weight * el_sf * mu_sf);

  auto nlep_req = df.constant(2);
  auto cut_2los = incl.filter<cut>("2l")(nlep_sel == nlep_req).filter<cut>("2los", [](const VecF& lep_charge){return lep_charge.at(0)+lep_charge.at(1)==0;})(lep_Q);
  auto cut_2ldf = cut_2los.channel<cut>("2ldf", [](const VecI& lep_type){return lep_type.at(0)+lep_type.at(1)==24;})(lep_type);
  auto cut_2lsf = cut_2los.channel<cut>("2lsf", [](const VecI& lep_type){return (lep_type.at(0)+lep_type.at(1)==22)||(lep_type.at(0)+lep_type.at(1)==26);})(lep_type);

  auto mll_cut = df.constant(80.0), met_cut = df.constant(30.0);
  auto cut_2ldf_sr = cut_2ldf.filter<cut>("sr")(mll > mll_cut);
  auto cut_2lsf_sr = cut_2lsf.filter<cut>("sr")(met > met_cut);
  auto cut_2ldf_cr = cut_2ldf.filter<cut>("cr")(mll > mll_cut);
  auto cut_2lsf_cr = cut_2lsf.filter<cut>("cr")(met > met_cut);

  auto pth_hist = df.book<Hist<1,float>>("pth",50,0,400).fill(pth);
  auto pth_2ldf_sr = pth_hist.at(cut_2ldf_sr);
  auto pth_2ldf_cr = pth_hist.at(cut_2ldf_cr);

  auto get_pt = df.define([](P4 const& p4){return p4.Pt();});
  auto l1pt = get_pt(l1p4);
  auto l2pt = get_pt(l2p4);
  auto l1n2pt_hists = df.book<Hist<1,float>>(std::string("l1n2pt"),50,0,200).fill(l1pt).fill(l2pt).at(cut_2los, cut_2lsf, cut_2ldf);
  auto pth_hists = df.book<Hist<1,float>>("pth",50,0,400).fill(pth).at(cut_2los, cut_2ldf, cut_2lsf);

  // pth_hists["2los"]->Draw(); gPad->Print("pth_2los.pdf");

  // // auto out_file = TFile::Open("hww_results.root","recreate");
  // // ana::output::dump<Folder>(pth_hists, *out_file, "ggf");
  // // delete out_file;

  auto mll_vars_2los = df.book<Hist<1,float>>("mll",50,0,100).fill(mll).at(cut_2los);
  auto mll_nom_2los = mll_vars_2los.get_nominal();
  auto mll_var_2los = mll_vars_2los["lp4_up"];
  mll_nom_2los->SetLineColor(kBlack); mll_nom_2los->Draw("hist");
  mll_var_2los->SetLineColor(kRed); mll_var_2los->Draw("E same");
  gPad->Print("mll_varied_2los.pdf");

  auto mll_vars_channels = df.book<Hist<1,float>>("mll",50,0,200).fill(mll).at(cut_2ldf, cut_2lsf);
  std::cout << mll_vars_channels.get_nominal()["2ldf"]->GetMean() << std::endl;
  std::cout << mll_vars_channels["lp4_up"]["2lsf"]->GetMean() << std::endl;

  return 0;
}

File: examples/tree_example.py
import os
import numpy as np

import ROOT
ana = ROOT.ana

ana.multithread.enable(2)
hww = ana.analysis['Tree']( ['hww.root'], 'mini' )
# hww.open( ['hww.root'], 'mini' )

mc_weight = hww.read['float']("mcWeight")
el_sf = hww.read['float']("scaleFactor_ELE")
mu_sf = hww.read['float']("scaleFactor_MUON")

lep_pt_MeV = hww.read['ROOT::RVec<float>']("lep_pt")
lep_eta = hww.read['ROOT::RVec<float>']("lep_eta")
lep_phi = hww.read['ROOT::RVec<float>']("lep_phi")
lep_E_MeV = hww.read['ROOT::RVec<float>']("lep_E")
lep_Q = hww.read['ROOT::RVec<float>']("lep_charge")
lep_type = hww.read['ROOT::RVec<unsigned int>']("lep_type")
met_MeV = hww.read['float']("met_et")
met_phi = hww.read['float']("met_phi")

print(mc_weight)

ROOT.gInterpreter.Declare('''
template <typename T> using RVec = ROOT::RVec<T>;
using RVecF = ROOT::RVec<float>;
using RVecD = ROOT::RVec<double>;
using RVecUI = ROOT::RVec<unsigned int>;
using TLV = TLorentzVector;
class NthP4 : public ana::column::definition<TLV(RVecD, RVecD, RVecD, RVecD)>
{

public:

  NthP4(unsigned int index) : 
    ana::column::definition<TLV(RVecD, RVecD, RVecD, RVecD)>(),
    m_index(index)
  {}

  virtual ~NthP4() = default;

  virtual TLV evaluate(ana::observable<RVecD> pt, ana::observable<RVecD> eta, ana::observable<RVecD> phi, ana::observable<RVecD> es) const override {
    TLV p4;
    p4.SetPtEtaPhiE(pt->at(m_index),eta->at(m_index),phi->at(m_index),es->at(m_index));
    return p4;
  }

protected:

  unsigned int m_index;

};

auto test_fn = std::function{[](int x){return x;}};
''')

l1p4 = hww.define['NthP4'](0)
print(l1p4)
l1p4 = l1p4.evaluate(lep_pt_MeV, lep_eta, lep_phi, lep_E_MeV)
print(l1p4)

lep_pt_GeV = hww.constant(10) + hww.constant(1000.0)
print(lep_pt_GeV)

print(type(ROOT.test_fn))
print(hww.calculate[type(ROOT.test_fn)](ROOT.test_fn))

mc_weighted = hww.filter['ana::selection::cut']("mcWeight")
print(mc_weighted)
mc_weighted = mc_weighted.evaluate(mc_weight)
print(mc_weighted)


hist = hww.book['Histogram<1,float>']('pt',100,0,100)
booked = hist.at(mc_weighted)

print(hist)

# lep_pt_hist = hww.book['Histogram<1,float>']("lep_pt",100,0,200).fill(lep_pt_MeV)

# print(lep_pt_hist)

# lep_pt_hist.result().Draw()
# ROOT.gPad.Print("lep_pt.pdf")
